src/services/rfqm_service.ts

```diff
@@ -0,0 +1,130 @@
+import { AssetSwapperContractAddresses, MarketOperation, ProtocolFeeUtils, QuoteRequestor } from '@0x/asset-swapper';
+import { BigNumber } from '@0x/utils';
+
+import { META_TX_WORKER_REGISTRY, RFQT_REQUEST_MAX_RESPONSE_MS } from '../config';
+import { NULL_ADDRESS, ONE_MINUTE_MS } from '../constants';
+
+export interface FetchIndicativeQuoteParams {
+    apiKey: string;
+    buyAmount?: BigNumber;
+    sellAmount?: BigNumber;
+    buyToken: string;
+    sellToken: string;
+    takerAddress?: string;
+}
+
+export interface FetchIndicativeQuoteResponse {
+    buyAmount: BigNumber;
+    buyTokenAddress: string;
+    price: BigNumber;
+    sellAmount: BigNumber;
+    sellTokenAddress: string;
+    gas: BigNumber;
+    allowanceTarget?: string;
+}
+
+export const EMPTY_QUOTE_RESPONSE = {
+    buyAmount: new BigNumber(0),
+    buyTokenAddress: '',
+    price: new BigNumber(0),
+    sellAmount: new BigNumber(0),
+    sellTokenAddress: '',
+    gas: new BigNumber(0),
+    allowanceTarget: NULL_ADDRESS,
+};
+
+const RFQM_DEFAULT_OPTS = {
+    takerAddress: NULL_ADDRESS,
+    txOrigin: META_TX_WORKER_REGISTRY || NULL_ADDRESS,
+    makerEndpointMaxResponseTimeMs: RFQT_REQUEST_MAX_RESPONSE_MS,
+    nativeExclusivelyRFQ: true,
+    altRfqAssetOfferings: {},
+    isLastLook: true,
+};
+
+/**
+ * RfqmService is the coordination layer for HTTP based RFQM flows
+ */
+export class RfqmService {
+    constructor(
+        private readonly _quoteRequestor: QuoteRequestor,
+        private readonly _protocolFeeUtils: ProtocolFeeUtils,
+        private readonly _contractAddresses: AssetSwapperContractAddresses,
+    ) {}
+
+    /**
+     * Fetch the best indicative quote available.
+     */
+    public async fetchIndicativeQuoteAsync(params: FetchIndicativeQuoteParams): Promise<FetchIndicativeQuoteResponse> {
+        const { sellAmount, buyAmount, sellToken, buyToken, apiKey } = params;
+
+        // Map params to the terminology of Quote Requestor
+        const isSelling = sellAmount !== undefined;
+        const marketOperation = isSelling ? MarketOperation.Sell : MarketOperation.Buy;
+        const assetFillAmount = isSelling ? sellAmount! : buyAmount!;
+        const makerToken = buyToken;
```

Comment:
Sorry @phil-ociraptor I forgot how this works, bear with me. Does this code already account for buys and sells? For example, `makerToken` is `ETH` if the user is selling `DAI` for `ETH`. However the `makerToken` will be `DAI` if the user is buying `DAI` for `ETH`.

Reply:
Resolved offline